File: Go/0008-string-to-integer-atoi.go
import (
	"math"
	"unicode"
)

func myAtoi(s string) int {
	i := 0
	n := len(s)
	sign := 1
	num := 0

	// 1. Skip leading whitespaces
	for i < n && s[i] == ' ' {
		i++
	}

	// 2. Handle optional sign
	if i < n && (s[i] == '+' || s[i] == '-') {
		if s[i] == '-' {
			sign = -1
		}
		i++
	}

	// 3. Convert digits and handle overflow
	for i < n && unicode.IsDigit(rune(s[i])) {
		digit := int(s[i] - '0')

		if num > (math.MaxInt32-digit)/10 {
			if sign == 1 {
				return math.MaxInt32
			}
			return math.MinInt32
		}

		num = num*10 + digit
		i++
	}

	return sign * num
}

File: Java/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int i = 0, n = s.length(), sign = 1, num = 0;

        // 1. Skip leading whitespaces
        while (i < n && s.charAt(i) == ' ') {
            i++;
        }

        // 2. Handle optional sign
        if (i < n && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            sign = s.charAt(i) == '-' ? -1 : 1;
            i++;
        }

        // 3. Convert digits and handle overflow
        while (i < n && Character.isDigit(s.charAt(i))) {
            int digit = s.charAt(i) - '0';

            // Check overflow
            if (num > (Integer.MAX_VALUE - digit) / 10) {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            num = num * 10 + digit;
            i++;
        }

        return sign * num;
    }
}
